src/gt4py/stencil_object.py

```diff
@@ -352,50 +420,25 @@ def _call_run(
             ValueError
                 If invalid data or inconsistent options are specified.
         """
-
         if exec_info is not None:
             exec_info["call_run_start_time"] = time.perf_counter()
 
-        domain_ndim = self.domain_info.ndim
-        origin = self._make_origin_dict(origin)
-        all_origin = origin.get("_all_", None)
-
-        # Set an appropriate origin for all fields
-        for name, field_info in self.field_info.items():
-            if field_info is not None:
-                assert name in field_args, f"Missing value for '{name}' field."
-                field_origin = origin.get(name, None)
-
-                if field_origin is not None:
-                    field_origin_ndim = len(field_origin)
-                    if field_origin_ndim != field_info.ndim:
-                        assert (
-                            field_origin_ndim == field_info.domain_ndim
-                        ), f"Invalid origin specification ({field_origin}) for '{name}' field."
-                        origin[name] = (*field_origin, *((0,) * len(field_info.data_dims)))
+        cache_key = _compute_cache_key(field_args, parameter_args, domain, origin)
+        if cache_key != self._last_key:
+            origin = self._normalize_origins(field_args, origin)
```

Comment:
I understand that checking if the current call matches the last one is the most common case, but I would add a dict here  `{cache_key: (origin, domain)}` (maybe limited in size?) to make sure that information from any previous call is reused, since the cost of computing the cache key is already much larger than a (small) dict lookup.